src/NHibernate/Engine/StatefulPersistenceContext.cs

```diff
@@ -801,10 +809,9 @@ public virtual object GetLoadedCollectionOwnerOrNull(IPersistentCollection colle
 			object loadedOwner = null;
 			// TODO: an alternative is to check if the owner has changed; if it hasn't then
 			// return collection.getOwner()
-			object entityId = GetLoadedCollectionOwnerIdOrNull(ce);
-			if (entityId != null)
+			if (ce.LoadedKey != null)
```

Comment:
Now that `GetCollectionOwner(object key, ...` correctly handles unique key, this small hack here attempting to resolve the id from the key is no more needed.
Moreover, all other calls to `GetCollectionOwner` were transmitting the collection key without this hack, so not always the id. (Due to bad naming, `DefaultInitializeCollectionEventListener.InitializeCollectionFromCache could let think this one was supplying the identifier, but indeed it is supplying the key too.)